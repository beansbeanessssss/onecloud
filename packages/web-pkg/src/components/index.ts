export * from './ContextActions'
export * from './sideBar'
export * from './Spaces'

export { default as AppLoadingSpinner } from './AppLoadingSpinner.vue'
export { default as AppTopBar } from './AppTopBar.vue'
export { default as BatchActions } from './BatchActions.vue'
export { default as ItemFilter } from './ItemFilter.vue'
export { default as LoadingIndicator } from './LoadingIndicator.vue'
export { default as NoContentMessage } from './NoContentMessage.vue'
export { default as QuotaSelect } from './QuotaSelect.vue'
export { default as SpaceQuota } from './SpaceQuota.vue'
