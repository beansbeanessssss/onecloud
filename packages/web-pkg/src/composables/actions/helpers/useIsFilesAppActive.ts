import { computed, unref } from 'vue'
import { activeApp, useRoute } from '../../router'

const isFilesAppActive = (activeApp: string): boolean => {
  // FIXME: we should use this constant but it somehow breaks the unit tests
  // return activeApp === FilesApp.appInfo.id
  return activeApp === 'files'
}

export const useIsFilesAppActive = () => {
  const currentRoute = useRoute()

  return computed(() => isFilesAppActive(activeApp(unref(currentRoute))))
}
