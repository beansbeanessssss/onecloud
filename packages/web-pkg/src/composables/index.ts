export * from './ability'
export * from './actions'
export * from './appDefaults'
export * from './archiverService'
export * from './authContext'
export * from './breadcrumbs'
export * from './capability'
export * from './clientService'
export * from './configuration'
export * from './download'
export * from './driveResolver'
export * from './fileListHeaderPosition'
export * from './filesList'
export * from './keyboardActions'
export * from './loadingService'
export * from './localStorage'
export * from './pagination'
export * from './passwordPolicyService'
export * from './piniaStores'
export * from './previewService'
export * from './router'
export * from './scrollTo'
export * from './search'
export * from './selection'
export * from './service'
export * from './sideBar'
export * from './sort'
export * from './spaces'
export * from './store'
export * from './upload'
export * from './viewMode'
