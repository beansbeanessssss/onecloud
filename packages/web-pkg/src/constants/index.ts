export * from './dav'
export * from './share'
