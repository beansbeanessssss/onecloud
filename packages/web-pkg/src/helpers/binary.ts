export const blobToArrayBuffer: (blob: Blob) => Promise<string | ArrayBuffer> = (blob: Blob) => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onloadend = () => resolve(reader.result)
    reader.onerror = (e) => reject(e)
    reader.readAsArrayBuffer(blob)
  })
}

export const canvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
  return new Promise((resolve) => canvas.toBlob(resolve))
}
