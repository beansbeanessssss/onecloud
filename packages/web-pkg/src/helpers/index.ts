export * from './cache'
export * from './contextMenuDropdown'
export * from './download'
export * from './datetime'
export * from './filesize'
export * from './locale'
