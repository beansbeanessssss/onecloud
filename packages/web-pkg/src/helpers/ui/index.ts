export * as resourceTable from './resourceTable'
