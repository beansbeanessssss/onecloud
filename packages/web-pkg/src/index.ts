export * from './components'
export * from './composables'
export * from './configuration'
export * from './errors'
export * from './helpers'
export * from './http'
export * from './observer'
export * from './services'
export * from './utils'
