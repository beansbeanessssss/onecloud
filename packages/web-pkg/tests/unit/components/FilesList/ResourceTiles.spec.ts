import { defaultPlugins, mount } from 'web-test-helpers'
import ResourceTiles from '../../../../src/components/FilesList/ResourceTiles.vue'
import { sortFields } from '../../../../src/helpers/ui/resourceTiles'
import { Resource } from '@ownclouders/web-client'
import { mock } from 'vitest-mock-extended'
import { computed } from 'vue'
import { extractDomSelector } from '@ownclouders/web-client/src/helpers'

vi.mock('../../../../src/composables/viewMode', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useTileSize: vi.fn().mockReturnValue({
    calculateTileSizePixels: vi.fn().mockImplementation((viewSize: number) => 100 * viewSize)
  })
}))

const mockUseEmbedMode = vi.fn().mockReturnValue({ isEnabled: computed(() => false) })
vi.mock('../../../../src/composables/embedMode', () => ({
  useEmbedMode: vi.fn().mockImplementation(() => mockUseEmbedMode())
}))

const router = {
  push: vi.fn(),
  afterEach: vi.fn(),
  currentRoute: {
    name: 'some-route-name',
    query: {},
    params: {
      driveAliasAndItem: ''
    }
  },
  resolve: (r) => {
    return { href: r.name }
  }
}

const spacesResources = [
  {
    id: '1',
    name: 'Space 1',
    path: '',
    type: 'space',
    isFolder: true,
    indicators: [],
    getDriveAliasAndItem: () => '1'
  },
  {
    id: '2',
    name: 'Space 2',
    path: '',
    type: 'space',
    isFolder: true,
    indicators: [],
    getDriveAliasAndItem: () => '2'
  }
]

const resources = [
  {
    id: 'forest',
    driveId: 'forest',
    name: 'forest.jpg',
    path: 'images/nature/forest.jpg',
    extension: 'jpg',
    thumbnail: 'https://cdn.pixabay.com/photo/2015/09/09/16/05/forest-931706_960_720.jpg',
    indicators: [],
    isFolder: false,
    type: 'file',
    tags: ['space', 'tag', 'moon'],
    size: '111000234',
    hidden: false,
    syncEnabled: true,
    outgoing: false,
    shareRoles: [],
    sharePermissions: [],
    shareTypes: [],
    canRename: vi.fn(),
    getDomSelector: () => extractDomSelector('forest')
  }
]

describe('ResourceTiles component', () => {
  const originalGetElementById = document.getElementById
  const originalGetComputedStyle = window.getComputedStyle
  beforeEach(() => {
    const mockElement = {
      clientWidth: 800
    }
    document.getElementById = vi.fn((id) => {
      if (id === 'tiles-view') {
        return mockElement
      }
      return originalGetElementById.call(document, id)
    })
    window.getComputedStyle = vi.fn().mockImplementation(() => {
      return {
        getPropertyValue: (propName) => {
          switch (propName) {
            case '--oc-size-tiles-default':
              return '9rem'
            case '--oc-size-tiles-resize-step':
              return '9rem'
            default:
              return originalGetComputedStyle(document.documentElement).getPropertyValue(propName)
          }
        },
        fontSize: '14px'
      }
    })
  })
  it('renders an array of spaces correctly', async () => {
    const { wrapper } = getWrapper({ resources: spacesResources })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a footer slot', () => {
    const { wrapper } = getWrapper({}, { footer: 'Hello, ResourceTiles footer!' })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits fileClick event upon click on tile', async () => {
    const { wrapper } = getWrapper({ resources: resources })
    await wrapper.find('.oc-tiles-item .oc-resource-name').trigger('click')
    expect(wrapper.emitted().fileClick[0][0].resources[0].name).toMatch('forest.jpg')
  })

  it('does not emit fileClick event upon click on tile when embed mode is enabled', async () => {
    mockUseEmbedMode.mockReturnValue({
      isEnabled: computed(() => true)
    })
    const { wrapper } = getWrapper({ resources: resources })
    await wrapper.find('.oc-tiles-item .oc-resource-name').trigger('click')
    expect(wrapper.emitted().fileClick).toBeUndefined()
  })

  it('emits update:selectedIds event on resource selection and sets the selection', () => {
    const { wrapper } = getWrapper({
      resources: spacesResources,
      selectedIds: [spacesResources[0].id]
    })
    wrapper.vm.toggleSelection(spacesResources[0])
    expect(
      wrapper.findComponent({ name: 'resource-tile' }).props('isResourceSelected')
    ).toBeTruthy()
    expect(wrapper.emitted('update:selectedIds')).toBeTruthy()
  })

  describe('sorting', () => {
    it('renders the label of the first sort field as default', () => {
      const { wrapper } = getWrapper({ sortFields })
      expect(wrapper.find('#oc-tiles-sort-btn').text()).toEqual(sortFields[0].label)
    })
    it('renders the label of the current sort field as default', () => {
      const sortField = sortFields[2]
      const { wrapper } = getWrapper({
        sortFields,
        sortBy: sortField.name,
        sortDir: sortField.sortDir
      })
      expect(wrapper.find('#oc-tiles-sort-btn').text()).toEqual(sortField.label)
    })
    it('emits the "sort"-event', async () => {
      const index = 2
      const { wrapper } = getWrapper({ sortFields })
      ;(wrapper.vm.$refs.sortDrop as any).tippy = { hide: vi.fn() }
      await wrapper.findAll('.oc-tiles-sort-list-item').at(index).trigger('click')
      expect(wrapper.emitted('sort')).toBeTruthy()
      expect(wrapper.emitted('sort')[0][0]).toEqual({
        sortBy: sortFields[index].name,
        sortDir: sortFields[index].sortDir
      })
    })
    describe('drag and drop', () => {
      it('emits the "update:selectedIds"-event on drag start', async () => {
        const { wrapper } = getWrapper()
        wrapper.vm.dragItem = mock<Resource>()
        await wrapper.vm.$nextTick()
        ;(wrapper.vm.$refs.ghostElementRef as any).$el = { style: {} }
        wrapper.vm.dragStart(mock<Resource>(), { dataTransfer: { setDragImage: vi.fn() } })
        expect(wrapper.emitted('update:selectedIds')).toBeDefined()
      })
      it('emits the "fileDropped"-event on resource drop', () => {
        const { wrapper } = getWrapper()
        wrapper.vm.fileDropped(mock<Resource>(), { dataTransfer: {} })
        expect(wrapper.emitted('fileDropped')).toBeDefined()
      })
    })
  })

  it.each([
    { viewSize: 1, expected: 'xlarge' },
    { viewSize: 2, expected: 'xlarge' },
    { viewSize: 3, expected: 'xxlarge' },
    { viewSize: 4, expected: 'xxlarge' },
    { viewSize: 5, expected: 'xxxlarge' },
    { viewSize: 6, expected: 'xxxlarge' }
  ])('passes the "viewSize" to the OcTile component', async (data) => {
    const { wrapper } = getWrapper({ resources: spacesResources, viewSize: data.viewSize })
    await wrapper.vm.$nextTick()
    expect(wrapper.findComponent({ name: 'resource-tile' }).props('resourceIconSize')).toEqual(
      data.expected
    )
  })

  function getWrapper(props = {}, slots = {}) {
    return {
      wrapper: mount(ResourceTiles, {
        props: {
          viewSize: 1,
          ...props
        },
        slots: {
          ...slots
        },
        global: {
          plugins: [...defaultPlugins()],
          mocks: {
            $route: router.currentRoute,
            $router: router
          },
          provide: {
            $router: router,
            $route: router.currentRoute
          }
        }
      })
    }
  }
})
