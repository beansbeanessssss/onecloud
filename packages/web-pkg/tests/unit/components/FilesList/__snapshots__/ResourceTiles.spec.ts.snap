// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceTiles component > renders a footer slot 1`] = `
"<div data-v-67d39700="" id="tiles-view" class="oc-px-m oc-pt-l">
  <!--v-if-->
  <oc-list data-v-67d39700="" class="oc-tiles oc-flex"> </oc-list>
  <!--v-if-->
  <div data-v-67d39700="" class="oc-tiles-footer">Hello, ResourceTiles footer!</div>
</div>"
`;

exports[`ResourceTiles component > renders an array of spaces correctly 1`] = `
"<div data-v-67d39700="" id="tiles-view" class="oc-px-m oc-pt-l">
  <!--v-if-->
  <oc-list data-v-67d39700="" class="oc-tiles oc-flex">
    <li data-v-67d39700="" class="oc-tiles-item has-item-context-menu">
      <resource-tile-stub data-v-67d39700="" resource="[object Object]" isresourceselected="false" isresourceclickable="true" isextensiondisplayed="true" resourceiconsize="xlarge" draggable="false" resourceroute="[object Object]"></resource-tile-stub>
    </li>
    <li data-v-67d39700="" class="oc-tiles-item has-item-context-menu">
      <resource-tile-stub data-v-67d39700="" resource="[object Object]" isresourceselected="false" isresourceclickable="true" isextensiondisplayed="true" resourceiconsize="xlarge" draggable="false" resourceroute="[object Object]"></resource-tile-stub>
    </li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
  </oc-list>
  <!--v-if-->
  <div data-v-67d39700="" class="oc-tiles-footer"></div>
</div>"
`;
